### PR TITLE
Contact Form Block: Add shortcode transform

### LINKED ISSUE #18210: Contact Form in Classic Editor Block converted to Shortcode Block
#### What I expected

When a contact form shortcode is in a classic editor block I expected it to become a contact form block when converted to a block.

#### What happened instead

It was converted to a shortcode block.

contact-form-shortcode

### PR DESCRIPTION
**Pull request by apeatling**
Fixes https://github.com/Automattic/jetpack/issues/18210

#### Changes proposed in this Pull Request:

* Add a block transform so that a contact form shortcode added to a classic block will transform into a contact form block. Previously this would transform only into a shortcode block.

| Before      | After |
| ----------- | ----------- |
|   |   |

#### Does this pull request change what data or activity we track or use?

No

#### Testing instructions:

**Note you will need to make a Gutenberg edit to test this:** I'm working on a core change that will let a raw transform override a shortcode transform when the priority is set higher.

* Edit `gutenberg/packages/block-library/src/shortcode/transforms.js` and completely remove the `from [ ... ]` entry.
* Rebuild gutenberg.
* Check out and build this branch and create a new post
* Copy this shortcode: https://cloudup.com/ceaECtt3X5Z
* Insert a classic block and paste the shortcode
* Click the "Convert to blocks" button
* Confirm that the shortcode is converted accurately into a form block.

#### Proposed changelog entry for your changes:

* No changelog needed.

### CODE REVIEW

**Comment by apeatling** [comment]
I have shortcode and nested shortcodes transforming into a proper contact form block. A few issues remain:

- [ ] * The core shortcode transform still takes precedence over this raw transform, even with a higher priority set.
- [ ] * The blockData needs to be reset after transform so that any further transforms don't inherit the data.
- [ ] * The subject and email address fields are not being parsed correctly.

**Comment by apeatling** [comment]
This is working now, see gif. There is an issue with Gutenberg where the [shortcode transform here](https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/shortcode/transforms.js) which has a priority of 20, cannot be overridden by a `raw` transform of priority 1. Until this is fixed this transform will not apply.

File: projects/plugins/jetpack/extensions/blocks/contact-form/transforms.js
Comment (glendaviesnz) [inline projects/plugins/jetpack/extensions/blocks/contact-form/transforms.js:29]: just a personal preference thing, but with this many switch options I find creating a lookup structure a slightly more concise, eg.:
```
const fieldTypes = {
	text: `${ prefix }/field-text`,
	url: `${ prefix }/field-text`,
	textarea: `${ prefix }/field-textarea`,
	radio: `${ prefix }/field-radio`,
	checkbox: `${ prefix }/field-checkbox`,
	select: `${ prefix }/field-select`,
	email: `${ prefix }/field-email`,
	name: `${ prefix }/field-name`,
	default: `${ prefix }/field-text`,
};
return fieldTypes[type] ? fieldTypes[type] : fieldTypes.default;
```

**Comment by glendaviesnz** [comment]
When I test this, with the transform removed from gutenberg, I just end up with one field in the block:

```
<!-- wp:jetpack/contact-form {"subject":"Form Submitted","to":"email@yourgroovydomain.com"} -->
<!-- wp:jetpack/field-name {"label":"Name","options":["Yes","No"],"placeholder":"Your Email Address"} /-->

<!-- wp:jetpack/button {"element":"button","text":"Contact Us"} /-->
<!-- /wp:jetpack/contact-form -->
```

which seems to have merged in the options from another field into the email address field

File: projects/plugins/jetpack/extensions/blocks/contact-form/transforms.js
Comment (glendaviesnz) [inline projects/plugins/jetpack/extensions/blocks/contact-form/transforms.js:9]: I would just spell these const names out in full `doubleQuotes` ... just a minor thing, but saves a few seconds of overhead working out what these abbreviations mean

File: projects/plugins/jetpack/extensions/blocks/contact-form/transforms.js
Comment (glendaviesnz) [inline projects/plugins/jetpack/extensions/blocks/contact-form/transforms.js:114]: I might be missing something, but I can't see how this loops through all the nested `contact-fields`, and when I test it does only transform one of them?

Comment (apeatling) [reply projects/plugins/jetpack/extensions/blocks/contact-form/transforms.js:114]: Do you have new lines after every contact field? I think that may be the issue -- not picking up each shortcode as a separate shortcode.

Comment (glendaviesnz) [reply projects/plugins/jetpack/extensions/blocks/contact-form/transforms.js:114]: I tried with or without a newline after each field and I get the same result.

Comment (apeatling) [reply projects/plugins/jetpack/extensions/blocks/contact-form/transforms.js:114]: hmm, that's odd. Are you using the shortcode example from the PR description?

Comment (apeatling) [reply projects/plugins/jetpack/extensions/blocks/contact-form/transforms.js:114]: It doesn't need to loop, it should be treating every contact field as a separate shortcode.

Comment (glendaviesnz) [reply projects/plugins/jetpack/extensions/blocks/contact-form/transforms.js:114]: yeh, just copying and pasting your code ... very strange ... I will do some debugging and try and work out what is going on - on the surface it appears to be just treating it as a single string, and doing a too greedy regex on it so attributes from different fields end up in the single block ... not sure why it does that for me and not you

**Comment by apeatling** [comment]
Fixed up those issues @glendaviesnz, ready for another review.

File: projects/plugins/jetpack/extensions/blocks/contact-form/transforms.js
Comment (glendaviesnz) [inline projects/plugins/jetpack/extensions/blocks/contact-form/transforms.js:10]: is it intentional that we want to check for ! `undefined` or `null` here? given how infrequently we end up using `!=` instead of `!==` it does look odd 😄

Comment (apeatling) [reply projects/plugins/jetpack/extensions/blocks/contact-form/transforms.js:10]: Yes, thanks. This was actually a copy and paste job :)

**Review by glendaviesnz — DISMISSED** [review]
This tests well for me now

File: projects/plugins/jetpack/extensions/blocks/contact-form/transforms.js
Comment (apeatling) [reply projects/plugins/jetpack/extensions/blocks/contact-form/transforms.js:10]: I'm also not used to these new ligatures I'm trying out 😂 

Screen Shot 2021-01-28 at 1 40 49 PM

**Comment by simison** [comment]
@glendaviesnz @apeatling  BTW since this is a transform, this seems safe to merge on simple sites even before it's shipping in Jetpack to Atomic sites?

**Review by samiff — DISMISSED** [review]
Tested well for me with the provided instructions and was able to convert the classic block w/shortcode --> form block without error.

**Review by jeherve — CHANGES_REQUESTED** [review]
The linting warnings are going to stop us from being able to merge this I'm afraid, it causes the tests to fail.

I also wonder if it would be beneficial to add some tests for `getAttributeValue`, to ensure we're catching different use-cases?

**Comment by simison** [comment]
Unrelated, but how come these are reported as warnings if they fail on CI? They should just be outright errors [in config](https://github.com/Automattic/jetpack/blob/fee24dc45c68329fd344493531a4cb16217efda7/.eslintrc.js#L134), methinks. :-) That'll help IDEs and commits fail on them, I think.

Could of course be missing something, in which case just ignore me. :-D

image

**Comment by aaronrobertshaw** [comment]
The escape characters are needed for the regex to function correctly. To fix this I've added comments to disable the `no-useless-escape` ESLint rule.

This still tests as advertised for me 👍

**Comment by andrewserong** [comment]
This is testing well for me, too!

**Review by jeherve — COMMENTED** [review]
I can't seem to get this to work, I must be missing something in my setup. All I get is this JavaScript error:

```
Uncaught TypeError: Cannot read property 'length' of null
    at transform (transforms.js:111)
```

**Comment by aaronrobertshaw** [comment]
@jeherve I had some trouble earlier today getting my environment working to test this but didn't come across that error. Were you using the same shortcode from the snippet in the testing instructions, or something else you could provide? 

I've just pulled the latest on this branch and it's tested ok for me again.

- pulled latest changes
- `yarn install && yarn build`
- made the edit to Gutenberg and rebuilt it
- started up the local dev environment
- created a new post, added classic block and pasted the shortcode into it before converting to blocks
- tried a range of shortcodes including some deliberately broken
- tried a few means of disabling the Gutenberg shortcode transform

Unfortunately, I couldn't replicate the error.

**Comment by jeherve** [comment]
> Were you using the same shortcode from the snippet in the testing instructions, or something else you could provide?

I tried with the provided shortcode as well as with this one (it has an invalid `submit_botton` attr):

```
[contact-form to="jeremy@myemailaddress.com" subject="A custom subject for example" submit_button="Click the button to send"][contact-field label="Enter your name below" type="name" required="1"][contact-field label="Email" type="email" required="1"][contact-field label="Website" type="url"][contact-field label="Comment" type="textarea" required="1"][contact-field label="What's your favourite colour?" type="radio" required="1" options="blue,yellow,green,red"][contact-field label="Pick the right answer" type="checkbox-multiple" required="1" options="Option A,Option A but with a twist,Option B (or is it C)?,Obiwan Kenobi"][/contact-form]
```

> made the edit to Gutenberg and rebuilt it

To clarify, I just set `transforms` to an empty object, should I have done things differently?

**Comment by aaronrobertshaw** [comment]
Thanks for the shortcode. 

I get the same error you reported using that. If I make each of the contact fields close with `/]` instead of just `]`. Then the shortcode gets converted without throwing the error. That said, it doesn't look like one of the fields displays correctly. I'm not very familiar with the contact form shortcodes but can take another look at this tomorrow.

The following snippet is the one used in the GIF below and has the fields ending in `/]`.

```
[contact-form to="jeremy@myemailaddress.com" subject="A custom subject for example" submit_button="Click the button to send"][contact-field label="Enter your name below" type="name" required="1" /][contact-field label="Email" type="email" required="1" /][contact-field label="Website" type="url" /][contact-field label="Comment" type="textarea" required="1" /][contact-field label="What's your favourite colour?" type="radio" required="1" options="blue,yellow,green,red" /][contact-field label="Pick the right answer" type="checkbox-multiple" required="1" options="Option A,Option A but with a twist,Option B (or is it C)?,Obiwan Kenobi" /][/contact-form]
```


ContactFormTransformsRetest

The [shortcode on cloudup](https://cloudup.com/ceaECtt3X5Z) still works for me and doesn't produce any errors.

> To clarify, I just set transforms to an empty object, should I have done things differently?

I tried with both an empty object and an empty `from` array. It didn't make any difference for me.

`const transforms = {};`

```
const transforms = {
	from: [],
};
```

**Comment by jeherve** [comment]
> If I make each of the contact fields close with /] instead of just ].

Oh, interesting. This was an old post of mine, so we definitely used that format at some point in the past.

**Comment by davemart-in** [comment]
Hey @apeatling just checking in on the status on this one since it's related to the list of issues from Matt's walkthrough in Jan.

**Comment by apeatling** [comment]
@davemart-in Getting back to this one today. It's going to need some core changes though so I don't expect to get it merged soon.

**Comment by apeatling** [comment]
> I get the same error you reported using that. If I make each of the contact fields close with /] instead of just ]. Then the shortcode gets converted without throwing the error.

The regex for parsing the contact-fields is expecting a closing slash. I've updated this to make the closing slash optional, and now both with and without closing slashes work. Nice find!

**Comment by apeatling** [comment]
> That said, it doesn't look like one of the fields displays correctly.

The code was missing a transform for the `checkbox-multiple` field. I've added this now and the field looks correct. 👍🏻

**Comment by apeatling** [comment]
@aaronrobertshaw and @jeherve I've fixed the found issues. I think this could get a final test and approval. It won't have any impact until the core change gets in, but it could merge before that and start working when that gets in?

**Review by aaronrobertshaw — APPROVED** [review]
LGTM 👍 

- [Shortcode](https://cloudup.com/ceaECtt3X5Z) in test instructions works
- [Old post shortcode ](https://github.com/Automattic/jetpack/pull/18454#issuecomment-775085619) without closing `/` works
- Same shortcode as above with `/` still works.

**Review by jeherve — APPROVED** [review]
This tests well for me, it should be good to merge. I don't know if you want to wait until the changes actually happen in Core? 

If you get some extra bandwidth, I still think it would be nice to get some tests to accompany this :)

**Comment by apeatling** [comment]
@jeherve We're going to spend the next sprint as a team adding tests to many of the Jetpack blocks, including this one. I'm going to merge this and use the upcoming work to add the tests. :+1:

**Comment by jeherve** [comment]
@apeatling Do you think it's possible to commit D55711-code now?

Thank you!

**Comment by apeatling** [comment]
r223925-wpcom.

**Pull request merged**
